psl-core/src/main/java/org/linqs/psl/reasoner/sgd/SGDReasoner.java

```diff
@@ -60,65 +78,86 @@ public double optimize(TermStore baseTermStore) {
             for (SGDObjectiveTerm term : termStore) {
-                movement += term.minimize(iteration, variableValues);
+                if (oldVariableValues != null) {
+                    objective += term.evaluate(oldVariableValues);
+                }
+
+                termCount++;
+                meanMovement += term.minimize(termStore.getVariableValues(), iteration, calculateAnnealedLearningRate(iteration),
+                        accumulatedGradientSquares, accumulatedGradientMean, accumulatedGradientVariance,
+                        adaGrad, adam, coordinateStep);
             }
 
-            if (variableValues.length != 0) {
-                movement /= variableValues.length;
-            }
+            termStore.iterationComplete();
 
-            long end = System.currentTimeMillis();
+            if (termCount != 0) {
+                meanMovement /= termCount;
+            }
 
-            oldObjective = objective;
-            objective = computeObjective(termStore);
-            totalTime += end - start;
+            converged = breakOptimization(objective, oldObjective, meanMovement, termCount);
 
-            if (log.isTraceEnabled()) {
-                log.trace("Iteration {} -- Objective: {}, Mean Movement: {}, Iteration Time: {}, Total Optimiztion Time: {}",
-                        iteration, objective, movement, (end - start), totalTime);
+            // Keep track of the old variables for a deferred objective computation.
+            if (oldVariableValues == null) {
+                oldVariableValues = Arrays.copyOf(termStore.getVariableValues(), termStore.getVariableValues().length);
+            } else {
+                System.arraycopy(termStore.getVariableValues(), 0, oldVariableValues, 0, oldVariableValues.length);
+                oldObjective = objective;
             }
 
-            iteration++;
-            termStore.iterationComplete();
+            long end = System.currentTimeMillis();
+            totalTime += end - start;
 
-            if (breakOptimization(iteration, objective, oldObjective, movement)) {
-                break;
+            if (iteration > 1) {
+                if (log.isTraceEnabled()) {
+                    log.trace("Iteration {} -- Objective: {}, Normalized Objective: {}, Iteration Time: {}, Total Optimization Time: {}",
+                            iteration - 1, objective, objective / termCount, (end - start), totalTime);
+                }
             }
         }
 
-        log.info("Optimization completed in {} iterations. Objective: {}, Total Optimiztion Time: {}",
-                iteration - 1, objective, totalTime);
-        log.debug("Optimized with {} variables and {} terms.", termStore.getNumVariables(), termStore.size());
+        objective = computeObjective(termStore);
+        log.info("Final Objective: {}, Final Normalized Objective: {}, Total Optimization Time: {}", objective, objective / termCount, totalTime);
```

Comment:
Again do we want normalized objective now?

Reply:
For logging, I think it is nice to report both the unnormalized and normalized objective values. This is useful for reporting convergence results and tuning hyperparameters.